maven-model/src/main/java/org/apache/maven/model/merge/ModelMerger.java

```diff
@@ -3018,4 +2618,298 @@ protected Object getExclusionKey( Exclusion exclusion )
         return exclusion;
     }
 
+    interface KeyComputer<T>
+    {
+        Object key( T t );
+    }
+
+    class DependencyKeyComputer implements KeyComputer<Dependency>
```

Comment:
private?
We can't make it static, can we?
If it is static we can have a single stateless instance

Reply:
The classes can't be make static and be able to reuse the methods in the enclosing class that actually compute the keys.   Would we require JDK8, a simple lambda `this::computeDependencyKey` could be used.  In the mean time, we need those non static inner classes.